stew/ranges/ptr_arith.nim

```diff
@@ -1,12 +1,17 @@
 proc baseAddr*[T](x: openarray[T]): pointer = cast[pointer](x)
 
-proc shift*(p: pointer, delta: int): pointer {.inline.} =
+# Please note that we use templates here on purpose.
+# As much as I believe in the power of optimizing compilers, it turned
+# out that the use of forced inlining with templates still creates a
+# significant difference in the release builds of nim-faststreams
+
+template shift*(p: pointer, delta: int): pointer =
```

Comment:
which is interesting - must come down to bad nim codegen because gcc has no issues inlining with this in general which means that simply by having a better cgen that doesn't add so much junk would make a significant difference.. I suspect this is in large parts why `nlvm` gets better perf as well in general: 
https://gcc.godbolt.org/z/MBNlDZ

Reply:
I think inlining is also affected by the overall "size" of your program (the compilers consider both the inlined function and where it's being inlined), so godbolt may not always be a completely accurate way to look at these things.

It's true that creating templates is a bit more dangerous, because they introduce the risk of accidental multiple evaluation of their arguments, but otherwise I think they bring several other small wins. When inlining happens in the front-end, this reduces the size of the generated code, which can also have a small positive effect on compilation speed (especially when the inline proc is generic). As an anecdote, I still remember how I reduced the compilation speed of Nim itself by 15% with a small tweak in the codegen that reduced the size of the generated code for line tracking.

Reply:
I guess this is also more problematic because of how LTO & `whole-program` is not being used by default - the matter of exported symbol names adds to the equation. that said, nim outputs a lot of redundant code like casts and multi-step initializations that probably add weight to the inlining cost function.. I wonder if one could offset that with more generous inlining cutoffs (`-finline-limit=n`).

the other issue I've run into with templates is that there are far more compiler bugs in type deduction and the like in there.